### PR TITLE
fix(query): align PIVOT BY semantics with bean-query

### LINKED ISSUE #1034: PIVOT robustness: empty-columns underflow + degenerate value==pivot case
Pre-existing risks in \`apply_pivot\` (\`crates/rustledger-query/src/executor/sort.rs\`), surfaced during review of #1033.

## 1. \`result.columns.len() - 1\` underflow potential

\`\`\`rust
let value_col_idx = result.columns.len() - 1;
\`\`\`

If \`result.columns\` is empty, this is \`0usize - 1\` — wraps and produces \`usize::MAX\`, then the filter \`*i != value_col_idx\` becomes a no-op silently.

In practice the path isn't reachable today: \`apply_pivot\` only runs when \`pivot_exprs\` is non-empty, and \`find_pivot_column\` errors before reaching this line for an empty \`result.columns\`. But the bound is implicit. Replace with \`checked_sub\` or an early return.

## 2. Degenerate case: \`value_col_idx == pivot_col_idx\`

For \`SELECT currency PIVOT BY currency\` (the value column IS the pivot column — single-column SELECT with PIVOT on that same column), the filter \`*i != pivot_col_idx && *i != value_col_idx\` excludes the index once, but the row builder treats them as separate exclusions. Behavior is probably fine in practice but undefined — no test, no documented contract.

Pin one of:
- Reject this query at parse/plan time (\"value column cannot be the pivot column\")
- Document the result shape and add a test

## Why filed as follow-up

Pre-existing — not introduced by #1023 or #1022. Out of scope for the PIVOT-precision PR. Filed here so the cleanup isn't lost.

## Acceptance criteria

- [ ] \`apply_pivot\` handles empty \`result.columns\` without underflow (early return or \`checked_sub\`).
- [ ] Decision pinned for \`value_col_idx == pivot_col_idx\` — either a parser/planner reject with a useful error, or a documented output shape with a unit test.

## Related

- #1023 / PR #1033 (where these were surfaced)

### PR DESCRIPTION
**Pull request by robcohen**
## Summary

Closes #1034. All four items in that issue collapse into one fix: align rledger's PIVOT BY with bean-query's semantics.

| #1034 item | Status |
|---|---|
| #1: `result.columns.len() - 1` underflow | already addressed by #1033; new `apply_pivot` doesn't need `value_col_idx` at all |
| #2: `value_col_idx == pivot_col_idx` | new `PivotSameColumn` error, bean-query message verbatim |
| #3: PIVOT BY semantic divergence | rewrote to take TWO columns, matching bean-query's `PIVOT BY <pivot_value_col>, <group_by_col>` |
| #4: PIVOT + ORDER BY hidden-column strip | dissolves: PIVOT now runs AFTER sort+strip, so strip operates on the pre-pivot shape where hidden cols are still trailing |

## What changed

**Syntax** (breaking — pre-existing PIVOT BY queries will error):
```
PIVOT BY <pivot_value_col>, <group_by_col>
```
First column's values become new column headers; second is the GROUP BY column kept as the row key.

**Three new validation errors** (matching bean-query's `_compile_pivot_by` wording verbatim for upstream parity):

| Error | Message |
|---|---|
| `PivotWrongArity(n)` | `PIVOT BY requires exactly two columns, got N` |
| `PivotSameColumn` | `the two PIVOT BY columns cannot be the same column` |
| `PivotSecondNotInGroupBy` | `the second PIVOT BY column must be a GROUP BY column` |

**Parser clause order** flipped: ORDER BY before PIVOT BY (matches bean-query grammar). Pre-fix rledger had PIVOT BY before ORDER BY, which is bean-query-incompatible at the grammar level.

**Execution pipeline** reordered:
- Pre-fix:  aggregate → PIVOT → sort → strip-hidden → LIMIT
- Post-fix: aggregate → sort → strip-hidden → PIVOT → LIMIT

PIVOT now operates on the visible-only, sorted shape. The strip step always sees hidden cols at trailing positions (pre-pivot), so it correctly removes them. Item #4 dissolves cleanly without needing to track hidden col indices through PIVOT's reshape.

**Rewritten `apply_pivot`**:
- Validates arity, distinctness, GROUP-BY membership before doing any work.
- Collects pivot values in first-seen order (preserves user's ORDER BY through to the new column-header order).
- Generalizes to multiple value columns: with exactly one (typical `SUM(number)` case), the pivot value alone is the new column header; with multiple, columns are qualified `<value_col_name> / <pivot_value>`.
- The `column_currency_hints` fallback from #1033 (precision via #1023) remains intact — pivoted column names are still currency strings.

## Test plan

New tests:
- `test_parse_pivot_by_two_columns` — bean-query-compatible form parses.
- `test_parse_pivot_by_one_column_parses_but_executes_with_arity_error` — pins the parser/executor split.
- `test_pivot_by_same_column_rejected` — item #2.
- `test_pivot_by_second_column_must_be_in_group_by` — new validation.
- `test_pivot_by_with_order_by_works` — regression for item #4 (visible-col case).
- `test_pivot_by_with_order_by_on_hidden_column_works` — the killer test for item #4. Pre-fix this query silently dropped the pivot values; post-fix the USD column is intact and the hidden `MIN(date)` is correctly stripped.

Updated tests: existing renderer e2e tests in `cmd::query::output` updated to use two-column form.

- [x] 511 query tests + 307 rustledger lib tests + 20 renderer tests pass.
- [x] `cargo clippy --all-features --all-targets -- -D warnings` clean.
- [x] `cargo fmt --all -- --check` clean.

## Migration impact

Any existing user PIVOT BY queries with one column will now error with `PivotWrongArity(1)`. Most queries migrate by adding the GROUP BY column to keep as the row key:

```sql
-- before
PIVOT BY currency

-- after
PIVOT BY currency, account
```

**Multi-column GROUP BY caveat.** Pre-fix rledger kept ALL non-pivot, non-value GROUP BY columns as a composite row key. The new bean-query-aligned form lets you keep only ONE:

```sql
-- pre-fix: rows keyed by (account, year) — both visible
SELECT account, year, currency, SUM(amount)
GROUP BY account, year, currency
PIVOT BY currency

-- post-fix: must pick ONE row key. account OR year, not both:
PIVOT BY currency, account   -- year drops out of the row key
PIVOT BY currency, year      -- account drops out
```

Bean-query has the same limitation (it's a property of the two-column form), so this is upstream-compatible behavior. Users who relied on the composite-key semantics need to either restructure (drop the extra dimension) or aggregate further.

**Implicit GROUP BY.** Queries that aggregate without an explicit `GROUP BY` clause (`SELECT SUM(amount) PIVOT BY currency, account`) are now rejected with the new `PivotWithoutGroupBy` error, distinct from `PivotSecondNotInGroupBy`.

The post-fix output shape is also bean-query-compatible (key column first, then pivot-value columns) — that's a shape change for PIVOT users, but the pre-fix shape was inconsistent in its own right (column header included the value column but rows didn't, fixed in #1033).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by robcohen — COMMENTED** [review]
Self-review pass. Five real concerns, two minor observations.

## Real concerns

**1. Migration is more invasive than the PR description claims.**

The PR body says "Migration is mechanical: add the GROUP BY column they want as the row key" with a one-column example. The actual migration story for **multi-column GROUP BY** is more limiting:

```sql
-- Pre-fix: rows keyed by (account, year) — both visible
SELECT account, year, currency, SUM(amount)
GROUP BY account, year, currency
PIVOT BY currency

-- Post-fix: must pick ONE row key. Can't preserve both account AND year:
PIVOT BY currency, account   -- year drops out of the row key
PIVOT BY currency, year      -- account drops out
```

Bean-query has the same limitation (it's a fundamental property of their two-column form), so this is upstream-compatible behavior — but users who relied on rledger's "keep all other GROUP BY cols as composite key" semantics need to either restructure their query or aggregate further. Worth being explicit in the PR body so users can plan around it.

**2. Hash-only dedup can collide silently.**

```rust
let mut seen: std::collections::HashSet<u64> = std::collections::HashSet::new();
for row in &result.rows {
    let v = row.get(pivot_value_col_idx).cloned().unwrap_or(Value::Null);
    let h = hash_single_value(&v);
    if seen.insert(h) { pivot_values.push(v); }
}
```

If two distinct `Value`s hash to the same `u64`, the second one is silently dropped from `pivot_values` and its data ends up under the first one's column header. Probability ≈ 2⁻⁶⁴ per pair, so never in practice — but the pattern is already used in the existing `pivot_lookup: HashMap<u64, &&Row>` (line 296), so this PR doesn't make it worse. Worth filing a follow-up to swap to `HashMap<Value, …>` with structural `Eq`/`PartialEq` (Value already implements them).

**3. Multi-value-column case not tested.**

The new layout-string logic explicitly designs for multiple value columns:

```rust
if value_col_idxs.len() == 1 {
    new_columns.push(pv_str);
} else {
    for &vci in &value_col_idxs {
        new_columns.push(format!("{} / {pv_str}", result.columns[vci]));
    }
}
```

But no test exercises the `> 1 value cols` branch. A query like `SELECT account, currency, SUM(number), COUNT(*) GROUP BY 1, 2 PIVOT BY currency, account` would hit the multi-value path with output columns like `["account", "SUM(number) / USD", "COUNT(*) / USD", "SUM(number) / EUR", ...]`. Either add a test or `#[cfg(test)] panic!()` the multi-value path until someone needs it (YAGNI — bean-query's multi-value behavior should be the reference if/when it's added).

**4. Implicit GROUP BY (no `GROUP BY` clause but aggregates in SELECT) + PIVOT errors with PivotSecondNotInGroupBy.**

```rust
let key_in_group_by = group_by.as_ref().is_some_and(|gb| { ... });
if !key_in_group_by { return Err(QueryError::PivotSecondNotInGroupBy); }
```

`group_by.as_ref()` is `None` when there's no explicit `GROUP BY` clause. The is_some_and short-circuits to `false`. Result: `PivotSecondNotInGroupBy` for ANY PIVOT on an implicit-grouping query.

This is probably correct (PIVOT BY without a key dimension is undefined), but it's not what the error message says. Either:
- Add a more specific error variant: `PivotRequiresExplicitGroupBy`.
- Document this case: implicit grouping is incompatible with PIVOT.
- Add a test pinning the current (somewhat misleading-message) behavior.

**5. `QueryError` is not `#[non_exhaustive]`.**

```rust
pub enum QueryError {
    ...
    PivotWrongArity(usize),
    PivotSameColumn,
    PivotSecondNotInGroupBy,
}
```

Downstream consumers that `match` exhaustively on `QueryError` will fail to compile after this PR. Pre-existing concern (the enum has had additions before), but each addition compounds it. If we want a stable public API, slap `#[non_exhaustive]` on the enum either in this PR or a follow-up.

## Minor observations

**6. `pivot_exprs.is_empty()` check is dead code today.**

```rust
if pivot_exprs.is_empty() { return Ok(result.clone()); }
```

The parser uses `at_least(1)` for `pivot_by_clause`, so `Option<Vec<Expr>>` is either `None` or `Some(vec)` with ≥ 1 element. The caller (`execute_select`) only calls `apply_pivot` inside `if let Some(pivot_exprs) = ...`. So `pivot_exprs.is_empty()` is unreachable in practice. Defensive-but-harmless. Could remove or convert to a `debug_assert!` for clarity.

**7. The default-sort-by-first-column block runs pre-strip, so it sorts on the original column layout.**

```rust
} else if has_grouping && !result.rows.is_empty() && !result.columns.is_empty() {
    let first_col = result.columns[0].clone();
    let default_order = vec![OrderSpec {
        expr: Expr::Column(first_col),
        direction: SortDirection::Asc,
    }];
    self.sort_results(&mut result, &default_order)?;
}
```

`result.columns[0]` at this point is the first VISIBLE select target (hidden cols are appended later — verified at execution.rs:65 where extended_targets is built). Correct as-is, but the invariant "hidden cols are appended AFTER visible" is implicit. Worth a one-line comment so a future refactor doesn't reorder them and silently break this default-sort path.

## Verdict

Concerns 1 and 4 are about clarity/correctness signaling and want either doc updates or tests. 2 and 5 are pre-existing code-quality issues that this PR puts in the spotlight; both could be follow-ups. 3 is "untested-by-design" — pin a `panic!()` or add a test. 6 and 7 are micro-comments.

Want me to land 1 (PR body update), 3 (multi-value test or panic), and 4 (test pinning implicit-grouping behavior) in a follow-up commit on this branch? 2 and 5 belong in their own follow-up issue. 6 and 7 can ride along.

**Review by robcohen — COMMENTED** [review]
Self-review pass v2 (after the 6813f317 follow-up). Three real items, two minor.

## Worth fixing

**1. Docstring lists 3 validation rules but there are 4 now.**

```rust
/// **Validation** (matches `_compile_pivot_by` in
/// `beanquery/compiler.py`):
/// - exactly two pivot columns (`PivotWrongArity` otherwise)
/// - the two columns must differ (`PivotSameColumn` otherwise)
/// - the second column must be a GROUP BY target
///   (`PivotSecondNotInGroupBy` otherwise)
```

I added `PivotWithoutGroupBy` but forgot to update the docstring. Should add a fourth bullet:

```
/// - PIVOT requires an explicit GROUP BY clause
///   (`PivotWithoutGroupBy` otherwise — distinct from
///   `PivotSecondNotInGroupBy`, which is for "GROUP BY exists but the
///   key column isn't in it")
```

Also worth noting: `PivotWithoutGroupBy` is rledger-specific — bean-query doesn't have it because their grammar requires the second pivot column to be in GROUP BY (and they don't reach apply_pivot when GROUP BY is absent). The docstring's "matches `_compile_pivot_by`" claim is now slightly overstated.

**2. Multi-row same (key, pivot_value) pair silently picks the first row.**

```rust
let matched = pivot_lookup.get(&pv_hash).and_then(|bucket| {
    bucket.iter().find(|row| ...)
});
```

`bucket.iter().find(...)` returns the FIRST match. If the input has two rows with the same (key, pivot_value) pair — say `(Assets, USD, 100)` and `(Assets, USD, 200)` — the second row's value silently disappears.

In practice this can't happen post-aggregation: `GROUP BY account, currency` produces one row per (account, currency) pair. But the function doesn't enforce that the input came from aggregation. A caller that constructs PIVOT input from non-aggregated data (or a future code path that bypasses the validator) would get silently-wrong output.

Mitigations to consider:
- Test: add `test_pivot_by_with_duplicate_key_pivot_pairs_uses_first` to pin the current behavior. Cheapest.
- Comment: clarify "expects one row per (key, pivot_value) pair — typically guaranteed by GROUP BY".
- `debug_assert!`: panic on the duplicate case. Catches programmer errors but ships fine.

I'd lean toward "comment + test pinning the behavior" — fix-it-if-it-becomes-a-problem.

**3. `Vec<&&Row>` is an awkward type, simplifies to `Vec<&Row>`.**

```rust
let mut pivot_lookup: HashMap<u64, Vec<&&Row>> = HashMap::new();
for row in &group_rows {
    pivot_lookup.entry(...).or_default().push(row);
}
```

`group_rows: Vec<&Row>`, so iterating `&group_rows` yields `&&Row`. The `&&Row` works but reads like a typo. Cleaner:

```rust
let mut pivot_lookup: HashMap<u64, Vec<&Row>> = HashMap::new();
for &row in &group_rows {
    pivot_lookup.entry(...).or_default().push(row);
}
```

Same behavior, one fewer borrow level. Pure readability win.

## Minor (not blockers)

**4. Untested edge cases.** Empty result with PIVOT, ALL-NULL pivot values, NULL key column values. Each produces defensible-but-not-pinned output. Adding 1–2 tests for the canary cases would prevent silent regressions, but YAGNI is also defensible since these don't occur in real queries.

**5. The `if value_col_idxs.len() == 1` branch in column-name construction is a small special case.** Without it, columns would always be `<value_col_name> / <pivot_value>`. With it, they're just `<pivot_value>`. The single-value case matches bean-query (which uses just the pivot value), so it's correct, but the asymmetry is worth a one-line comment explaining the bean-query alignment.

## Verdict

Items 1–3 are easy fixes (docstring update, comment + test, type cleanup). 4 and 5 are documentation polish.

Want me to land 1+2+3 in a follow-up commit on this branch?

File: crates/rustledger-query/src/executor/sort.rs [hunk 1]
```diff
@@ -86,121 +86,148 @@ impl Executor<'_> {
...
-            .map(|(_, c)| c.clone())
+        // Collect unique pivot values, preserving the row-order they
+        // appear in (which post-sort means the user's ORDER BY drives
+        // the new column order). Stable dedup via a HashSet sidecar.
+        let mut pivot_values: Vec<Value> = Vec::new();
+        let mut seen: std::collections::HashSet<u64> = std::collections::HashSet::new();
+        for row in &result.rows {
+            let v = row.get(pivot_value_col_idx).cloned().unwrap_or(Value::Null);
+            let h = hash_single_value(&v);
+            if seen.insert(h) {
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-query/src/executor/sort.rs]: Already fixed in commit `6813f317` (one of the self-review-driven follow-ups in this PR). The new pivot value collection uses linear-scan dedup with structural `PartialEq`:

```rust
let mut pivot_values: Vec<Value> = Vec::new();
for row in &result.rows {
    let v = row.get(pivot_value_col_idx).cloned().unwrap_or(Value::Null);
    if !pivot_values.contains(&v) {
        pivot_values.push(v);
    }
}
```

`Value` impls `Eq` but not `Hash` (some inner types — `Decimal`, `Inventory` — don't impl `Hash`), so a wrapper type was the alternative. Linear-scan is simpler and pivot values are typically small (handful of currencies). Documented in an inline comment.

File: crates/rustledger-query/src/executor/sort.rs [hunk 2]
```diff
@@ -86,121 +86,148 @@ impl Executor<'_> {
...
-                .join("|");
-            groups.entry(key).or_default().push(row);
+            let key = row.get(key_col_idx).cloned().unwrap_or(Value::Null);
+            let h = hash_single_value(&key);
+            if let Some(&idx) = group_index.get(&h) {
+                groups[idx].1.push(row);
+            } else {
+                group_index.insert(h, groups.len());
+                groups.push((key, vec![row]));
+            }
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-query/src/executor/sort.rs:244]: Already fixed in commit `6813f317`. `group_index` is now `HashMap<u64, Vec<usize>>` — hash bucket + structural `==` inside the bucket:

```rust
let bucket = group_index.entry(h).or_default();
let existing = bucket.iter().find(|&&idx| groups[idx].0 == key).copied();
if let Some(idx) = existing {
    groups[idx].1.push(row);
} else {
    bucket.push(groups.len());
    groups.push((key, vec![row]));
}
```

Same effect as `HashMap<Value, …>` would give but works around the missing `Hash` impl on `Value`.

File: crates/rustledger-query/src/executor/sort.rs [hunk 3]
```diff
@@ -86,121 +86,148 @@ impl Executor<'_> {
...
+            let pivot_lookup: HashMap<u64, &&Row> = group_rows
                 .iter()
-                .enumerate()
-                .filter_map(|(idx, row)| {
-                    row.get(pivot_col_idx).map(|v| (hash_single_value(v), idx))
+                .map(|row| {
+                    let pv = row.get(pivot_value_col_idx).cloned().unwrap_or(Value::Null);
+                    (hash_single_value(&pv), row)
                 })
                 .collect();
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-query/src/executor/sort.rs]: Already fixed in commit `6813f317` (and further cleaned up in `6fe8ea02` from `Vec<&&Row>` → `Vec<&Row>`). `pivot_lookup` now stores all rows per hash bucket and disambiguates structurally on lookup:

```rust
let matched = pivot_lookup.get(&pv_hash).and_then(|bucket| {
    bucket
        .iter()
        .find(|row| row.get(pivot_value_col_idx).is_some_and(|cell| cell == pv))
});
```

Same hash-bucket + structural-verify pattern as `group_index`.

File: crates/rustledger-query/src/executor/sort.rs [hunk 4]
```diff
@@ -86,121 +86,148 @@ impl Executor<'_> {
...
+        // The grammar allows `PIVOT BY` without `GROUP BY`, but the
+        // semantics of the second column require it to identify a
+        // grouping key — otherwise the pivoted output rows wouldn't
+        // have a stable identity. Resolve each GROUP BY expression to
+        // its result column index and check membership.
+        let key_in_group_by = group_by.as_ref().is_some_and(|gb| {
+            gb.iter()
+                .filter_map(|expr| self.find_pivot_column(result, expr).ok())
+                .any(|idx| idx == key_col_idx)
+        });
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-query/src/executor/sort.rs]: Real concern, fixed in commit `59247857`. The root cause was `find_pivot_column`'s `_` arm rejecting ALL non-Column non-Literal exprs:

```rust
_ => Err(QueryError::Evaluation("PIVOT BY must reference a column name or index"))
```

Meaning `PIVOT BY YEAR(date)` would error AND `GROUP BY YEAR(date)` would silently filter out via `.ok()` in the membership check.

Added an explicit `Expr::Function` arm that mirrors `sort_results`'s ORDER BY two-tier resolution:
1. Try the function name (e.g. column `YEAR` for `YEAR(date)`)
2. Fall back to the full expression string

The `_` arm now also handles binary-op / other complex exprs via to_string match — same convention as the hidden-column alias path.

After this fix, `.filter_map(...ok())` in the membership check is correct: it only skips genuinely-unresolvable exprs, not silent failures from the lint resolver.

File: crates/rustledger-query/tests/bql_integration_test.rs [hunk 5]
```diff
@@ -1655,16 +1655,131 @@ fn test_having_filters_all() {
...
+    let directives = make_test_directives();
+    let result = execute_query(
+        "SELECT account, currency, SUM(number) GROUP BY 1, 2 \
+         ORDER BY account PIVOT BY currency, account",
+        &directives,
+    );
+    // Smoke check: produced a non-empty result with at least the
+    // account-key column and one pivoted currency column.
+    assert!(!result.columns.is_empty());
+    assert!(result.columns.iter().any(|c| c == "account"));
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-query/tests/bql_integration_test.rs]: Real concern, fixed in commit `59247857`. The test now has four assertions instead of two:

```rust
// 1. The key column survives
assert!(result.columns.iter().any(|c| c == "account"), ...);
// 2. At least one pivoted column appears (proves PIVOT ran)
assert!(result.columns.iter().any(|c| c == "USD"), ...);
// 3. Original currency column is GONE (its values became headers)
assert!(!result.columns.iter().any(|c| c == "currency"), ...);
// 4. Original value column is GONE (values moved into pivoted cells)
assert!(!result.columns.iter().any(|c| c == "SUM(number)" || c == "SUM"), ...);
```

A regression where PIVOT silently no-ops would now fail at assertions 2-4.
